data_steward/tools/purge_hpo_data.py

```diff
@@ -0,0 +1,100 @@
+"""
+Empty all CDM tables associated with one or more HPO sites from an EHR dataset
+
+This can be used to exclude sites' data from the EHR dataset before 
+generating the unioned dataset which will be used for the CDR.
+
+NOTE: This does not affect non-clinical tables (e.g. achilles_*)
+"""
+import logging
+from typing import List
+
+from google.cloud import bigquery
+
+from utils import bq
+import bq_utils
+import resources
+
+LOGGER = logging.getLogger(__name__)
+LIST_TABLES_MAX_RESULTS = 5000
+DELETE_QUERY_TPL = bq.JINJA_ENV.from_string("""
+{%- for table in tables_to_purge -%}
+DELETE FROM `{{table.project}}.{{table.dataset_id}}.{{table.table_id}}` WHERE 1=1;
+{%- endfor -%}
+""")
+
+
+def _filter_hpo_tables(tables: List[bigquery.table.TableListItem],
+                       hpo_id: str) -> List[bigquery.table.TableListItem]:
+    """
+    Given a list of tables get those associated with an HPO submission
+
+    :param tables: list of tables to filter
+    :param hpo_id: identifies the HPO
+    :return: list of tables assoicated with the HPO
+    """
+    expected_tables = [
+        bq_utils.get_table_id(hpo_id, table) for table in resources.CDM_TABLES
+    ]
+    return [table for table in tables if table.table_id in expected_tables]
+
+
+def purge_hpo_data(client: bigquery.Client, dataset: bigquery.DatasetReference,
+                   hpo_ids: List[str]) -> bigquery.QueryJob:
+    """
+    Empty all CDM tables associated with one or more HPO sites
+    
+    :param client: Active bigquery client object 
+    :param dataset: the dataset to purge data from
+    :param hpo_ids: Identifies the HPO sites whose data should be purged
+    :return: Query job associated with removing all the records
+    :raises RuntimeError if CDM tables associated with a site are not found in the dataset
+    """
+    LOGGER.debug(
+        f'purge_hpo_data called with dataset={dataset.dataset_id} and hpo_ids={hpo_ids}'
+    )
+    all_tables = list(
+        client.list_tables(dataset=dataset,
+                           max_results=LIST_TABLES_MAX_RESULTS))
+    tables_to_purge = []
+    for hpo_id in hpo_ids:
+        hpo_tables = _filter_hpo_tables(all_tables, hpo_id)
+        if not hpo_tables:
+            raise RuntimeError(
+                f'No tables found for {hpo_id} in dataset {dataset.dataset_id}. '
+                f'Ensure the specified arguments are correct.')
+        tables_to_purge.extend(hpo_tables)
+    script = DELETE_QUERY_TPL.render(tables_to_purge=tables_to_purge)
+    LOGGER.debug(f'purge_hpo_data about to start script:\n {script}')
+    return client.query(script)
+
+
+def main(credentials, project_id, dataset_id, hpo_ids: List[str]):
+    client = bigquery.Client.from_service_account_json(credentials)
```

Comment:
I agree- this script is simple and it'd be a good place to start using impersonation. I'm not sure of a few parts and I would prefer to discuss these with you beforehand. Could we look at impersonation separate from this PR?

For example, I'm not sure whether/how to get stored user credentials to use as source credentials. My impression is that we'd want the end-user (e.g. mark.velez) credentials to be used to acquire impersonated credentials for the service account (i.e. cdr-ops); theoretically this would enable us to more easily track down the individual associated with API activity. Currently we'd have to peruse BigQuery log events associated with the activity in question to find the SA key and then dig into the service account key logs to find out who created that key.

Reply:
@mark-velez I used impersonation in the copy_prototype.py script I created and gave to Nishanth to run.  He stripped the service account impersonation part out of that script, but you can still see the original script on DC-1030.  My example used my personal account service account key, because the script requires that level of credentials.  That key file had to be passed to access the impersonated service account key file.  So, I would still need service account credentials, but they would be my credentials, not the credentials of the service account running the code.  This is an example of my log file via [Google Logging](https://console.cloud.google.com/logs/viewer?project=aou-res-curation-test&minLogLevel=0&expandAll=false&timestamp=2020-10-05T17%3A20%3A44.366000000Z&customFacets&limitCustomFacetWidth=true&dateRangeStart=2020-09-21T13%3A20%3A44.000Z&dateRangeEnd=2020-10-05T17%3A20%3A44.618Z&interval=CUSTOM&resource=bigquery_resource&scrollTimestamp=2020-09-29T22%3A07%3A39.310000000Z&advancedFilter=resource.type%3D%22bigquery_resource%22%0Aresource.labels.project_id%3D%22aou-res-curation-test%22%0Atimestamp%3D%222020-09-29T22%3A07%3A36.790000000Z%22%0AinsertId%3D%22i1dg89d15rl%22).

Reply:
I think we can continue exploring service account impersonation separately from this PR.